lambda/matching.ml

```diff
@@ -2742,23 +2803,23 @@ let complete_pats_constrs = function
     to jump to in case of failure of elementary tests
 *)
 
+let comp_exit ctx def =
+  let i, jumps =
+    match Default_environment.pop def with
+    | Some ((i, _), _) -> i, Jumps.singleton i ctx
+    | None -> Default_environment.final_exit def, Jumps.empty Partial
+  in (Lstaticraise (i, []), jumps)
+
 let mk_failaction_neg partial ctx def =
   debugf
     "@,@[<v 2>COMBINE (mk_failaction_neg %a)@]"
     pp_partial partial
   ;
   match partial with
-  | Partial -> (
-      match Default_environment.pop def with
-      | Some ((idef, _), _) ->
-          (Some (Lstaticraise (idef, [])), Jumps.singleton idef ctx)
-      | None ->
-          (* Act as Total, this means
-             If no appropriate default matrix exists,
-             then this switch cannot fail *)
-          (None, Jumps.empty)
```

Comment:
This odd case of an empty default environment in`Partial` mode has its behavior change slightly in the new definition below. In the new code, this will generate a jump to the final exit, while the previous code would not. This is one case where the compilation behavior is in fact modified, and the new behavior is slightly worse than before.

(A high-level summary: we are going in the direction of trusting the type-checker totality information less, to avoid generating incorrect code when it is wrong. But this may introduce some cases where we will generate slightly worse code than if we had trusted it blindly. This here is one such case, but it is very rare, as explained below.)

I believe that this case basically never happens. When this current code was first written, there was an `assert false` in this case because the author (Luc Maranget) believed that it could not happen. But then Gilles Peskine in 2004 found an odd example that happens to trigger precisely this scenario. Here is the minimal repro case:

```ocaml
(*
  By Gilles Peskine, compilation raised some assert false i make_failactionneg
*)
type bg = [
  | `False
  | `True
]

type vg =
  | A
  | B
  | U of int
  | V of int
  | W of int

type tg = {
    v : vg;
    x : bg;
  }

let predg x = true

let rec gilles o = match o with
  | (* clause 1 *) {v = (U _ | V _); x = `False} when predg o -> 1
  | (* clause 2 *) {v = (A | B) ; x = `False}
  | (* clause 3 *) {v = (U _ | V _ | W _); x = `False}
    -> 2
  | (* clause 4 *) {v = _; x = `True} -> 3
```

A rough summary of what happens is as follows:
1. the whole program is Total, so in trunk there would be no final handler case added in the default environment -- compilation starts with an empty default environment
2. the compiler decides to first check clauses 1, 2, 4 together, and then clause 3 in a separate submatrix; the compilation of the submatrix 1,2,4 is in Partial mode (as it is followed by another submatrix), with an exit for clause 3 in the default environment
3. the compiler splits 1,2,4 into 4 followed by 1,2
4. when compiling 1,2 together, the compiler first generates a switch on column `v`, and then jumps to sub-matrices that match on column `x` (the details of why this happens are in a documentation comment for the `precompile_or` function)
5. when it compiles the second column of clause 2, it does this in a restricted default environment where all patterns incompatible with `v = (A | B)` have been filtered out (this is the call to `Default_environment.pop_compat` in `precompile_or`); this default environment is empty, so the compilation of the second column happens in Partial mode but in an empty default environment

In this case, our PR (or any approach to include a final-exit case in the Total case as well) will generate slightly worse code: the compilation of the `x` column of clause 2 will include a failure case if the value is different from `False` jumping to the final exit, and the whole pattern will thus include an (unnecessary) final exit handler.

Note that all three of (a) a guard, (b) several or-patterns in the same column, and (c) polymorphic variants at a closed type are necessary in this example. Without (b), we don't get the call to `pop_compat`, without (a), we don't get the split in step 2., so the compilation happens in a Total context. Without (c) (with a normal variant instead of a polymorphic variant), we don't get a call to `mk_failaction_neg`, we get `mk_failaction_pos` which is aware than there are no other cases that can match and does not generate a final exit.

Reply:
> I believe that this case basically never happens.

Famous last words! I believe that @ncik-roberts found examples of this pattern in the wild in https://github.com/ocaml/ocaml/pull/13338#issuecomment-2259647128 .